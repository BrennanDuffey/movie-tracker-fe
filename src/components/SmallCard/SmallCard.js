import React from 'react';
import { connect } from 'react-redux';
import { toggleFavorite } from '../../actions';

const SmallCard = ({ title, id, genres, summary, release, backdrop, poster, rating, loading, user, isFavorite, toggleFavorite }) => {
  
  // let genresList;

  // if (genres.length > 0) {
  //   let lastGenre = genres.pop();
  //   let genresString = `${genres.name.join(', ')} , and ${lastGenre.name}`;
  //   genresList = <p>Genres: {genresString}</p>
  // }
  const handleFavorite = (e) => {
    e.preventDefault();
    if (!isFavorite) {
      addFavorite();
    } else {
      removeFavorite();
    }
  }

  const removeFavorite = async () => {
    const url = `http://localhost:3000/api/users/${user.id}/favorites/${id}`;
    const init = { method:'DELETE' }
    try {
      const response = await fetch(url, init);
      if (!response.ok) {
        throw Error(response.text)
      }
      toggleFavorite(id);
    } catch {

    }
  }

  const addFavorite = async () => {
    const url = 'http://localhost:3000/api/users/favorites/new';
    const init = {
      method: 'POST',
      headers: {'Content-Type': 'application/json'},
      body: JSON.stringify({
        title: title,
        user_id: user.id,
        movie_id: id,
        overview: summary,
        poster_path: poster,
        release_date: release,
        vote_average: rating
      })
    }
    try {
      const response = await fetch(url, init);
      if(!response.ok) {
        throw Error(response.statusText)
      }
      toggleFavorite(id)
    } catch {
      // make something pop up to sat please sign in to favorite
    }
  }

  return (
    <article className="SmallCard">
      <div className="image--container">
        {poster && <img src={poster} alt={title} />}
      </div>
      <div className="content">
        {title && <h3>{title}</h3>}
        {summary && 
          <span className="sub-content">
            <div className="sub-border"></div>
            <p>{summary.slice(0,50)}</p>
            <p className="showmore">... (Show More)</p>
          </span>
        }

      </div>
      <div className="rating">
        {rating && <h4>{rating}</h4>}
      </div>
      <button onClick={handleFavorite}>Favorite</button>
    </article>
    
  )
}

const mapStateToProps = (state) => ({
  movies: state.movies,
  user: state.user
});

const mapDispatchToProps = (dispatch) => ({
  toggleFavorite: (id) => dispatch(toggleFavorite(id))
  // will use this to update movies in state so they become favorited on click
  // will need to update favorited movies on component did mount as well when app is first loaded
})

export default connect(mapStateToProps, mapDispatchToProps)(SmallCard);