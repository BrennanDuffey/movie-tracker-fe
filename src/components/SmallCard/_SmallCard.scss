.CardContainer {

  a {
    margin: min(20px) min(20px);
  }

  .rating {
    z-index: 5;
  }
}

.fave-true {
  color: rgb(232, 181, 27);

  &:hover {
    color: rgb(91, 91, 91);
  }
}