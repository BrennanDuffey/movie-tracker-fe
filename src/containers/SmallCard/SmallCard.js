import React, {Component} from 'react';
import { connect } from 'react-redux';
import {Route, NavLink} from 'react-router-dom';
import { toggleFavorite, grabCurrMovie } from '../../actions';
import BigCard from '../BigCard/BigCard';

export class SmallCard extends Component {
  constructor(props) {
    super();
  }
  
  // let genresList;

  // if (genres.length > 0) {
  //   let lastGenre = genres.pop();
  //   let genresString = `${genres.name.join(', ')} , and ${lastGenre.name}`;
  //   genresList = <p>Genres: {genresString}</p>
  // }
  handleFavorite = (e) => {
    e.preventDefault();
    console.log(this.props.movie)
    let { isFavorite } = this.props.movie;
    if (!isFavorite) {
      this.addFavorite();
    } else {
      this.removeFavorite();
    }
  }

  removeFavorite = async () => {
    let { user, toggleFavorite } = this.props;
    let { id } = this.props.movie
    const url = `http://localhost:3000/api/users/${user.id}/favorites/${id}`;
    const init = { method:'DELETE' }
    try {
      const response = await fetch(url, init);
      if (!response.ok) {
        throw Error(response.text)
      }
      toggleFavorite(id);
    } catch {

    }
  }

  addFavorite = async () => {
    let { user, toggleFavorite } = this.props;
    let {title, id, summary, poster, release, rating} = this.props.movie;
    const url = 'http://localhost:3000/api/users/favorites/new';
    const init = {
      method: 'POST',
      headers: {'Content-Type': 'application/json'},
      body: JSON.stringify({
        title: title,
        user_id: user.id,
        movie_id: id,
        overview: summary,
        poster_path: poster,
        release_date: release,
        vote_average: rating
      })
    }
    try {
      const response = await fetch(url, init);
      if(!response.ok) {
        throw Error(response.statusText)
      }
      toggleFavorite(id)
    } catch {
      // make something pop up to sat please sign in to favorite
    }
  }

  render() {
    let { movie } = this.props
    let { title, id, summary, poster, rating, isFavorite } = this.props.movie;
    let favClass = `${isFavorite}`;

    return (
      <NavLink className="small-link" to={`/movie/${id}`}>
        <article
          className="SmallCard"
          onClick={() => this.props.grabCurrMovie(movie)}
        >
          <div className="image--container">
            {poster && <img src={poster} alt={title} />}
          </div>
          <div className="content">
            {title && <h3>{title}</h3>}
            {summary && (
              <span className="sub-content">
                <div className="sub-border" />
                <p>{summary.slice(0, 50)}</p>
                <p className="showmore">... (Show More)</p>
              </span>
            )}
          </div>
          <div className="rating">{rating && <h4>{rating}</h4>}</div>
          <button className="faveBtn" onClick={this.handleFavorite}>
            <i className={`fas fa-star fave-${favClass}`} />
          </button>
          <Route
            path={`/movie/`}
            render={({ match }) => (
              <BigCard
                
                movie={this.props.movie}
              />
            )}
          />
        </article>
      </NavLink>
    );
  }
}

const mapStateToProps = (state) => ({
  movies: state.movies,
  user: state.user
});

const mapDispatchToProps = (dispatch) => ({
  toggleFavorite: (id) => dispatch(toggleFavorite(id)),
  // will use this to update movies in state so they become favorited on click
  // will need to update favorited movies on component did mount as well when app is first loaded
  grabCurrMovie: (movie)=> dispatch(grabCurrMovie(movie))
})

export default connect(mapStateToProps, mapDispatchToProps)(SmallCard);