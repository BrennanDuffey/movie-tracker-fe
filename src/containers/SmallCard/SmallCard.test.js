import {SmallCard} from './SmallCard';
import React from 'react';
import { shallow } from 'enzyme';

let mockMovie = {
  title: 'movie',
  poster: 'http',
  backdrop: 'http',
  summary: 'words',
  genres: ['action'],
  isFavorite: false,
  release: 2000,
  rating: 9.9,
  id: 69
}

describe('SmallCard', () => {
  let wrapper;

  beforeEach(()=>{
    wrapper = shallow(
      <SmallCard movie={mockMovie} key={mockMovie.id}  />
    );
  });

  it('should render to the dom correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });
});