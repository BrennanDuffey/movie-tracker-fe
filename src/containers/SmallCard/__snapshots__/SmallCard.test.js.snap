// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SmallCard should render to the dom correctly 1`] = `
<NavLink
  className="small-link"
  to="/movie/69"
>
  <article
    className="SmallCard"
    onClick={[Function]}
  >
    <div
      className="image--container"
    >
      <img
        alt="movie"
        src="http"
      />
    </div>
    <div
      className="content"
    >
      <h3>
        movie
      </h3>
      <span
        className="sub-content"
      >
        <div
          className="sub-border"
        />
        <p>
          words
        </p>
        <p
          className="showmore"
        >
          ... (Show More)
        </p>
      </span>
    </div>
    <div
      className="rating"
    >
      <h4>
        9.9
      </h4>
    </div>
    <button
      className="faveBtn"
      onClick={[Function]}
    >
      <i
        className="fas fa-star fave-false"
      />
    </button>
    <Route
      path="/movie/"
      render={[Function]}
    />
  </article>
</NavLink>
`;
